src/apprt/gtk-ng/class/command_palette.zig

```diff
@@ -0,0 +1,426 @@
+const Application = @import("application.zig").Application;
+const Window = @import("window.zig").Window;
+const Config = @import("config.zig").Config;
+
+const log = std.log.scoped(.gtk_ghostty_command_palette);
+
+pub const CommandPalette = extern struct {
+    pub inline fn isSupported() bool {
+        return comptime adw_version.supportsDialogs();
+    }
+
+    const Self = @This();
+    parent_instance: Parent,
+    pub const Parent = adw.Bin;
+    pub const getGObjectType = gobject.ext.defineClass(Self, .{
+        .name = "GhosttyCommandPalette",
+        .instanceInit = &init,
+        .classInit = &Class.init,
+        .parent_class = &Class.parent,
+        .private = .{ .Type = Private, .offset = &Private.offset },
+    });
+
+    pub const properties = struct {
+        pub const config = struct {
+            pub const name = "config";
+            const impl = gobject.ext.defineProperty(
+                name,
+                Self,
+                ?*Config,
+                .{
+                    .nick = "Config",
+                    .blurb = "The configuration that this command palette is using.",
+                    .accessor = gobject.ext.privateFieldAccessor(
+                        Self,
+                        Private,
+                        &Private.offset,
+                        "config",
+                    ),
+                },
+            );
+        };
+    };
+
+    const Private = struct {
+        /// The configuration that this command palette is using.
+        config: ?*Config = null,
+
+        /// The window that the command palette is attached to.
+        window: ?*Window = null,
+
+        arena: ArenaAllocator,
+
+        /// The dialog object containing the palette UI.
+        dialog: *adw.Dialog,
+
+        /// The search input text field.
+        search: *gtk.SearchEntry,
+
+        /// The view containing each result row.
+        view: *gtk.ListView,
+
+        /// The model that provides filtered data for the view to display.
+        model: *gtk.SingleSelection,
+
+        /// The list that serves as the data source of the model.
+        /// This is where all command data is ultimately stored.
+        source: *gio.ListStore,
+
+        pub var offset: c_int = 0;
+    };
+
+    /// Create a new instance of the command palette, connected to the given
+    /// window, and using the given config. The config should already have had
+    /// it's reference count increased and we take ownership of it. The instance
+    /// returned will be floating.
+    pub fn new(window: *Window, config: *Config) ?*Self {
+        if (comptime !Self.isSupported()) return null;
+
+        const self = gobject.ext.newInstance(Self, .{});
+
+        const app = Application.default();
+        const priv = self.private();
+
+        // The config that we will use to begin with. It should already have
+        // had it's reference count increased, and we take ownership of it.
+        priv.config = config;
```

Comment:
I found its cleaner to make sure these are properties with proper setter/getters and setup initialization in `init` rather than `new`. That's how we do config in other classes.

Reply:
I found that would leak a reference to the `config` because if you pass the config to `gobject.ext.newInstance` it would increment the references of `config`. Let me try again as I was doing a lot of this late at night.

Reply:
May need to update this with the custom private ref-counted accessor as discussed on Discord but I've moved all the logic to `.init`.

Reply:
Rebased with the new custom getter/setter for ref counted objects.